core/services/relay/evm/chain_writer.go

```diff
@@ -156,7 +169,38 @@ func (w *chainWriter) GetTransactionStatus(ctx context.Context, transactionID uu
 }
 
 func (w *chainWriter) GetFeeComponents(ctx context.Context) (*commontypes.ChainFeeComponents, error) {
-	return nil, fmt.Errorf("not implemented")
+	if w.ge.estimator == nil {
+		return nil, fmt.Errorf("not implemented")
+	}
+
+	gasPriceWei, _, err := w.ge.estimator.GetFee(ctx, nil, 0, assets.NewWei(w.ge.maxGasPrice))
+	if err != nil {
+		return nil, err
+	}
+	// Use legacy if no dynamic is available.
+	gasPrice := gasPriceWei.Legacy.ToInt()
+	if gasPriceWei.DynamicFeeCap != nil {
+		gasPrice = gasPriceWei.DynamicFeeCap.ToInt()
+	}
+	if gasPrice == nil {
+		return nil, fmt.Errorf("missing gas price %+v", gasPriceWei)
```

Comment:
Why use a `%+v` here? That seems unusual

Reply:
This was copied directly from CCIP's [gas estimation](https://github.com/smartcontractkit/ccip/blob/4a6f68e052f1e55596b1f8238f23806a7128c541/core/services/ocr2/plugins/ccip/prices/exec_price_estimator.go#L39), since they are the main consumer of this function. I suppose they wanted to include the names of the fields in the error message.

Reply:
Does it make sense to you why the internals of the type should be exposed? It does not make sense to me and it inherently carries risk.

None of the field in this type are even called 'price':
```go
type EvmFee struct {
	// legacy fees
	Legacy *assets.Wei

	// dynamic/EIP1559 fees
	DynamicFeeCap *assets.Wei
	DynamicTipCap *assets.Wei
}
```

IMHO this error should be telling the user what happened instead of making them figure it for themselves. `gasPrice == nil` means that we failed to derive the gas price from this other type, but the reason is that both `Legacy` and `DynamicFeeCap` were nil, so that is what the error should highlight.

Reply:
Updated the error message to be more specific